lean4.html.markdown

```diff
@@ -0,0 +1,587 @@
+    _       = a*a + b*a + (a*b + b*b) := by repeat rw [Nat.add_mul]
+    _       = a*a + b*a + a*b + b*b   := by rw [← Nat.add_assoc]
+    _       = a*a + a*b + a*b + b*b   := by rw [Nat.mul_comm b a]
+    _       = a^2 + a*b + a*b + b*b   := by rw [← Nat.pow_two a]
+    _       = a^2 + a*b + a*b + b^2   := by rw [← Nat.pow_two b]
+    _       = a^2 + (a*b + a*b) + b^2 := by rw [Nat.add_assoc (a^2)]
+    _       = a^2 + 2*(a*b) + b^2     := by rw [← Nat.two_mul (a*b)]
+    _       = a^2 + 2*a*b + b^2       := by rw [Nat.mul_assoc 2 a b]
+
+/-
+Let us now prove more "realistic" theorems. Those involving logical connectives.
+
+First, we define even and odd numbers.
+-/
+def Even (n : Nat) := ∃ k, n = 2*k
+def Odd  (n : Nat) := ∃ k, n = 2*k + 1
+
+/-
+To prove an existential, we can provide specific values if we know them.
+-/
+theorem zero_even : Even 0 :=
+  let h : 0 = 2 * 0 := Eq.symm (Nat.mul_zero 2)
+  Exists.intro 0 h
+/-
+`Exists.intro v h` proves `∃ x, p x` by substituting `x` by `v` and using the
+proof `h` for `p v`.
+-/
+
+/-
+Now, we will see how to use hypothesis that are existentials to prove
+conclusions that are existentials.
+-/
+theorem even_mul_even_is_even' (hn : Even n) (hm : Even m) : Even (n*m) :=
+  Exists.elim hn (fun k1 hk1 =>
+    Exists.elim hm (fun k2 hk2 =>
+      Exists.intro (k1 * ( 2 * k2)) (
+        calc
+          n*m = (2 * k1) * (2 * k2) := by rw [hk1, hk2]
+          _   = 2 * (k1 * (2 * k2)) := by rw [Nat.mul_assoc]
+      )
+    )
+  )
+
+/-
+The same theorem, proved using tactics.
+-/
+theorem even_mul_even_is_even (hn : Even n) (hm : Even m) : Even (n*m) := by
+  let ⟨k1, hk1⟩ := hn
+  let ⟨k2, hk2⟩ := hm
+  use k1 * (2 * k2)
+  calc
+    n*m = (2 * k1) * (2 * k2) := by rw [hk1, hk2]
+    _   = 2 * (k1 * (2 * k2)) := by rw [Nat.mul_assoc]
+
+/-
+Let us work with implications.
+-/
+theorem succ_of_even_is_odd' : Even n → Odd (n+1) :=
+  fun hn =>
+    let ⟨k, hk⟩ := hn
+    Exists.intro k (
+      calc
+        n + 1 = 2 * k + 1 := by rw [hk]
+    )
+/-
+To prove an implication `p → q`, you have to write a function that takes a proof
+of `p` and construct a proof of `q`.
+
+Here, `pn` is proof of `Even n := ∃ k, n = 2 *k`. Eliminating the existential
+gets us `k` and a proof `hk` of `n = 2 * k`.
+
+Now, we have to introduce the existential `∃ k, n + 1 = 2 * k + 1`. This `k` is
+the same as `k` for `n`. And, the equation is proved by a simple calculation
+that substitutes `2 * k` for `n`, which is allowed by `hk`.
+-/
+
+/-
+Same theorem, now using tactics.
+-/
+theorem succ_of_even_is_odd : Even n → Odd (n+1) := by
+  intro hn
+  let ⟨k, hk⟩ :=  hn
+  use k
+  rw [hk]
+
+/-
+The following theorem can be proved similarly.
+
+We will use this theorem later.
+-/
+theorem  succ_of_odd_is_even : Odd n → Even (n+1) := sorry
+
+/-
+We can use implications by applying them.
+-/
+example : Odd 1 := by
+  apply succ_of_even_is_odd
+  exact zero_even
+/-
+The two new tactics are:
```

Comment:
`use` is a stronger apply (it's a `refine`) and has been used harum scarum. Might deserve the same treatment as the following two.